app/services/foreman_bootdisk/iso_generator.rb

```diff
@@ -76,12 +99,42 @@ def self.generate(opts = {})
 
         if opts[:ipxe]
           ipxe_source_file = File.join(Setting[:bootdisk_ipxe_dir], 'ipxe.lkrn')
-          raise(Foreman::Exception, N_('Please ensure the ipxe-bootimgs package is installed.')) unless File.exist?(ipxe_source_file)
+          raise Foreman::Exception.new(N_('Please ensure the ipxe-bootimgs package is installed.')) unless File.exist?(ipxe_source_file)
 
           FileUtils.cp(ipxe_source_file, File.join(wd, 'build', 'ipxe'))
           File.open(File.join(wd, 'build', 'script'), 'w') { |file| file.write(opts[:ipxe]) }
         end
 
+        if opts[:grub]
+          FileUtils.mkdir_p(File.join(wd, 'build', 'EFI', 'BOOT'))
+          # a copy when using CDROM directly
+          File.open(File.join(wd, 'build', 'EFI', 'BOOT', 'grub.cfg'), 'w') { |file| file.write(opts[:grub]) }
+          # a copy when using USB/HDD (copied into EFI image as well)
+          File.open(File.join(wd, 'build', 'grub-hdd.cfg'), 'w') do |f|
+            # set root to (cd0) or similar when booting via disk
+            f.write("search --file /ISOLINUX.BIN --set root\n")
+            f.write(opts[:grub])
+          end
+          efibootimg = File.join(wd, 'build', 'efiboot.img')
+          system_or_exception("truncate -s 4M #{efibootimg}", N_('Creating new image failed, install truncate utility'))
+          system_or_exception("mkfs.msdos #{efibootimg}", N_('Failed to format the ESP image via mkfs.msdos'))
+          system_or_exception("mmd -i #{efibootimg} '::/EFI'", N_('Failed to create a directory within the ESP image'))
+          system_or_exception("mmd -i #{efibootimg} '::/EFI/BOOT'", N_('Failed to create a directory within the ESP image'))
+          {
+            File.join(wd, 'build', 'grub-hdd.cfg').to_s => 'GRUB.CFG',
+            File.join(Setting[:bootdisk_grub2_dir], 'grubx64.efi').to_s => 'GRUBX64.EFI',
+            File.join(Setting[:bootdisk_grub2_dir], 'shimx64.efi').to_s => 'BOOTX64.EFI',
+          }.each do |src, dest|
+            raise(Foreman::Exception.new(N_('Ensure %{path} contains grubx64.efi and shimx64.efi: install TFTP module or copy those files into /var/lib/foreman/bootdisk (Grub2 directory setting)'), path: src)) unless File.exist?(src)
+            raise(Foreman::Exception.new(N_('Unable to mcopy %{path}'), path: src)) unless system("mcopy -m -i #{efibootimg} '#{src}' '#{"::/EFI/BOOT/#{dest}"}'")
+          end
+          mkiso_args = "-eltorito-alt-boot -e efiboot.img -no-emul-boot"
+          isohybrod_args = ["--uefi"]
+        else
+          mkiso_args = ''
+          isohybrod_args = []
+        end
```

Comment:
This damn block took me two weeks of finding out how things work, emails with grub developers and countless failed boots... :beer: